pkg/cmd/admin/node/node.go

```diff
@@ -0,0 +1,399 @@
+`
+
+func NewCommandManageNode(f *clientcmd.Factory, commandName, fullName string, out io.Writer) *cobra.Command {
+	cfg := &NodeConfig{}
+
+	cmd := &cobra.Command{
+		Use:   commandName,
+		Short: "Manage node operations: schedulable, evacuate, list-pods",
+		Long:  fmt.Sprintf(manageNodeLongDesc, fullName),
+		Run: func(c *cobra.Command, args []string) {
+
+			defaultNamespace, err := f.DefaultNamespace()
+			kcmdutil.CheckErr(err)
+
+			_, kc, err := f.Clients()
+			kcmdutil.CheckErr(err)
+
+			mapper, typer := f.Object()
+
+			cfg.DefaultNamespace = defaultNamespace
+			cfg.Kclient = kc
+			cfg.Writer = out
+			cfg.Mapper = mapper
+			cfg.Typer = typer
+			cfg.RESTClientFactory = f.Factory.RESTClient
+			cfg.Printer = f.Printer
+			cfg.Operations = []string{}
+			cfg.NodeNames = []string{}
+
+			if c.Flag("schedulable").Changed {
+				cfg.Operations = append(cfg.Operations, "schedulable")
+			}
+			if c.Flag("evacuate").Changed {
+				cfg.Operations = append(cfg.Operations, "evacuate")
+			}
+			if c.Flag("list-pods").Changed {
+				cfg.Operations = append(cfg.Operations, "list-pods")
+			}
+
+			if len(args) != 0 {
+				cfg.NodeNames = append(cfg.NodeNames, args...)
+			}
+
+			errList := []error{}
+			if errList = cfg.ValidateParams(); len(errList) != 0 {
+				err := fmt.Errorf("%v", errList)
+				kcmdutil.CheckErr(kcmdutil.UsageError(c, err.Error()))
+			}
+
+			err = cfg.RunManageNode()
+			kcmdutil.CheckErr(err)
+		},
+	}
+	flags := cmd.Flags()
+
+	flags.BoolVar(&cfg.Schedulable, "schedulable", false, "Control pod schedulability on the node.")
+	flags.BoolVar(&cfg.Evacuate, "evacuate", false, "Migrate all/selected pods on the node.")
+	flags.BoolVar(&cfg.ListPods, "list-pods", false, "List all/selected pods on the node.")
+
+	flags.BoolVar(&cfg.DryRun, "dry-run", false, "Show pods that will be migrated. Optional param for --evacuate")
+	flags.Var(&cfg.PodSelector, "pod-selector", "Label selector to filter pods on the node. Optional parameters for --evacuate or --list-pods")
+	flags.Var(&cfg.Selector, "selector", "Label selector to filter nodes. Either pass one/more nodes as arguments or use this node selector")
+
+	return cmd
+}
+
+func (cfg *NodeConfig) ValidateParams() []error {
+	errList := []error{}
+	if cfg.Selector.Provided() {
+		if _, err := labels.Parse(cfg.Selector.Value()); err != nil {
+			errList = append(errList, errors.New("--selector=<node_selector> must be a valid label selector"))
+		}
+		if len(cfg.NodeNames) != 0 {
+			errList = append(errList, errors.New("either specify --selector=<node_selector> or nodes but not both"))
+		}
+	} else if len(cfg.NodeNames) == 0 {
+		errList = append(errList, errors.New("either specify --selector=<node_selector> or nodes"))
+	}
+
+	numOps := len(cfg.Operations)
+	if numOps == 0 {
+		errList = append(errList, errors.New("must provide a node operation. Supported operations: --schedulable, --evacuate and --list-pods"))
+	} else if numOps != 1 {
```

Comment:
> On May 12, 2015, at 7:58 PM, David Eads <notifications@github.com> wrote:
> 
> In pkg/cmd/admin/node/node.go:
> 
> > +	errList := []error{}
> > +	if cfg.Selector.Provided() {
> > +		if _, err := labels.Parse(cfg.Selector.Value()); err != nil {
> > +			errList = append(errList, errors.New("--selector=<node_selector> must be a valid label selector"))
> > +		}
> > +		if len(cfg.NodeNames) != 0 {
> > +			errList = append(errList, errors.New("either specify --selector=<node_selector> or nodes but not both"))
> > +		}
> > +	} else if len(cfg.NodeNames) == 0 {
> > +		errList = append(errList, errors.New("either specify --selector=<node_selector> or nodes"))
> > +	}
> > +
> > +	numOps := len(cfg.Operations)
> > +	if numOps == 0 {
> > +		errList = append(errList, errors.New("must provide a node operation. Supported operations: --schedulable, --evacuate and --list-pods"))
> > +	} else if numOps != 1 {
> Clayton recommended not to explode manage-node with many subcommands
> 
> Even without subcommands, modelling it that way in the code will make it much easier to describe an options (NodeConfig) object that makes sense. One for each of the operations will make intent clear in code, since only one operation will actually be performed.
> 
Yeah, my concern with subcommands is that admins will be operating on this command frequently and shouldn't have to read three help pages (and it's a "first touch" command for most admins).  Agree with David structuring the code as three separate distinct flows will improve maintainability and preserve the ability to add more flows in the future (or split into subcommands as we get more powerful).
> —
> Reply to this email directly or view it on GitHub.
>